### ISSUE TITLE
Cannot read property 'scrollTop' of null

### ISSUE THREAD

**Issue by malinduDeAlwis90** [mdbootstrap/perfect-scrollbar#835]
Hi,
My versions are:
"@angular/core": "7.0.2",
"ngx-perfect-scrollbar": "^6.3.1",
"swiper": "^4.4.1",

In my project I'm using perfect-scrollbar together with swiper.
Everything works fine when we do swipe the tabs.
But when we touch the screen and hold for a little time and then swipe will gives the following error

![image](https://user-images.githubusercontent.com/22950451/49078308-b2306400-f263-11e8-930e-e9a4547c17d2.png)
This occurs because at this point the setInterval below is never cleared.

perfect-scrollbar.esm.js - line 1045

```function touchEnd() {
    if (i.settings.swipeEasing) {
      clearInterval(easingLoop);
      easingLoop = setInterval(function() {
        if (i.isInitialized) {
          clearInterval(easingLoop);
          return;
        }

        if (!speed.x && !speed.y) {
          clearInterval(easingLoop);
          return;
        }

        if (Math.abs(speed.x) < 0.01 && Math.abs(speed.y) < 0.01) {
          clearInterval(easingLoop);
          return;
        }

        applyTouchMove(speed.x * 30, speed.y * 30);

        speed.x *= 0.8;
        speed.y *= 0.8;
      }, 10);
    }
  }
```

If there's a way to override this method we can fix it.

Any help?
Thanks,
Malindu

**Comment by olaf89** [comment]
Im also having alot of this issues. Would be awsome to get it fixed

**Comment by zecar** [comment]
no fixes yet? I'm having this issue too

**Comment by karthik02318** [comment]
Facing the same issue. I had added hammer event to track the swipe event in the same element used for scroll causes this issue. if checking element is not null should fix this issue

`function touchEnd() {
    if (i.settings.swipeEasing && i.element) {
      clearInterval(easingLoop);
      easingLoop = setInterval(function() {
        if (i.isInitialized) {
          clearInterval(easingLoop);
          return;
        }

        if (!speed.x && !speed.y) {
          clearInterval(easingLoop);
          return;
        }

        if (Math.abs(speed.x) < 0.01 && Math.abs(speed.y) < 0.01) {
          clearInterval(easingLoop);
          return;
        }

        applyTouchMove(speed.x * 30, speed.y * 30);

        speed.x *= 0.8;
        speed.y *= 0.8;
      }, 10);
    }
  }`

**Comment by anhhh11** [comment]
Set ps.isInitialized = true go to other screen in Angular (using scrollBar.directiveRef.ps(); scrollBar as PerfectScrollbarComponent )
It trigger i.isInitialized and interval is clear; no more error.
It's just workaround

**Comment by hyperknot** [comment]
I'm getting Sentry reports for this. The critical line is this:
https://github.com/mdbootstrap/perfect-scrollbar/blob/daeeddf5972c44a960f1d244a4b9719cb7c3d0b7/src/update-geometry.js#L8

Where element can be null. Do you think adding `if (!element) return` would fix this issue? I don't know the internals of perfect-scrollbar, but I believe if element is null then update-geometry shouldn't run on it.

**Comment by kamaladenalhomsi** [comment]
me too having the same issue

**Comment by rahuls360** [comment]
Same as @hyperknot.
I wonder if this issue crashes the app for the users.

**Comment by pentolbakso** [comment]
Yeah still happening. Some users report it crashes the app.

**Comment by captain-melanie** [comment]
For people who use the react version, [react-perfect-scrollbar](https://github.com/goldenyz/react-perfect-scrollbar), here's a hotfix.

<pre>
import React, { useEffect } from 'react';
import PerfectScrollbar from 'react-perfect-scrollbar';

export const WrappedPerfectScrollbar = PerfectScrollbarWithHotfix =>
    React.forwardRef((props, ref) => {
        useEffect(() => {
            return () => {
                // To clear the scheduler at perfect-scrollbar/src/handlers/touch.js:176
                // bc PerfectScrollbar might be destroyed before the timer is up which led to null i.element
                if (ref && ref.current && ref.current._ps) {
                    ref.current._ps.isInitialized = true;
                }
            };
        });
        return <PerfectScrollbarWithHotfix {...props} ref={ref} />;
    });

export const PerfectScrollbarWithHotfix = WrappedPerfectScrollbar(
    PerfectScrollbar
);
</pre>

usage:
<pre>
import { PerfectScrollbarWithHotfix as PerfectScrollbar } from 'PATH_TO_THE_HOTFIX';
</pre>

**Comment by pentolbakso** [comment]
Hi @captain-melanie 

I'm trying to apply your fix. But got this error:

React Hook "useEffect" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function  react-hooks/rules-of-hooks

**Comment by captain-melanie** [comment]
Hi @pentolbakso , it's likely that you called the hotfix component inside a hook. here's my usage of the component:

<pre>
import { PerfectScrollbarWithHotfix as PerfectScrollbar } from 'PATH_TO_THE_HOTFIX';

const SomeComp = props => {
    // some constants
    const PerfectScrollbarRef = useRef(null);
    // some constants

    return (
        // rendering components 
        < PerfectScrollbar ref={PerfectScrollbarRef}>
        // rendering components 
    }
}
</pre>

**Comment by riz007** [comment]
@captain-melanie I am having this issue in vue-perfect-scrollbar https://github.com/lecion/vue-perfect-scrollbar/issues/48
Is there any hotfix for this?

**Comment by captain-melanie** [comment]
@riz007 hi, unfortunately I don't have a hotfix for the vue version. The idea is to find a way to stop the scheduler https://github.com/mdbootstrap/perfect-scrollbar/blob/master/src/handlers/touch.js#L176 right before the component is unmounted. You may want to use the [vue unmount hook](https://v3.vuejs.org/api/options-lifecycle-hooks.html#unmounted): https://v3.vuejs.org/api/options-lifecycle-hooks.html#unmounted   
You wanted to have a reference pointing to the scrollbar variable (in my React version, it's `ref.current._ps.isInitialized`) and then reference to the `isInitialized` variable, set it to true to stop the scheduler.

**Comment by riz007** [comment]
@captain-melanie I am using Vue 2 and it doesn't have the `unmount` hook. However, I tried with the beforeDestroy hook instead and foudn that the `ps` value is always null here: 
![image](https://user-images.githubusercontent.com/12785997/100322515-f68f9f80-2ff6-11eb-9cab-ca06325f96bc.png)
basically it's set to null here in [vue-perfect-scrollbar's index ](https://github.com/lecion/vue-perfect-scrollbar/blob/master/index.vue)file.
So I cannot access to `ps.isInitialized`
Any other workarounds?

**Comment by guigonc** [comment]
@smolenski-mikolaj Would it be possible to do a release of this package with this fix?

**Comment by ahmed2m** [comment]
Just an update for any one subscribed to this thread, the PR was merged and published to npm version 1.5.1

**Comment by smolenski-mikolaj** [comment]
The PR was published. I'm closing the issue.
